### ISSUE TITLE
Periodic prune of let's encrypt (certbot) cert files

### ISSUE THREAD

**Issue by drmrbrewer** [caprover/caprover#934]
The number of `.pem` files in `/captain/data/letencrypt/etc/keys` is very large and getting larger over time.  At present I have 54979 of them, taking up 218 MB of space:

    # du -hs /captain/data/letencrypt/etc/keys
    218M    /captain/data/letencrypt/etc/keys

Apparently it might be possible to have CapRover initiate a cron-like job to clear unnecessary cert files away, since the let's encrypt client (certbot) doesn't have a native "prune" command... see related issue [here](https://github.com/caprover/caprover/issues/187#issuecomment-742489061).

**Comment by githubsaturn** [comment]
Looks like we can safely remove `csr` and `keys` directories which are consuming a lot of space

https://github.com/appsembler/roles/pull/61
https://github.com/certbot/certbot/issues/4634
https://github.com/certbot/certbot/issues/4635

**Comment by githubsaturn** [comment]
Feel free to take a back up of your system and try this out. If it works without any issues, we can incorporate this into CapRover so that it runs automatically.

**Comment by drmrbrewer** [comment]
Remove those directories completely?  Or just files older than 90 days?  Six months?  It's not clear to me from those linked posts.

**Comment by githubsaturn** [comment]
From those issues, it looks like those are just duplicated files: https://community.letsencrypt.org/t/remove-all-generated-csr-to-free-disk-space/78237/9

**Comment by drmrbrewer** [comment]
So after removing the files, do I need to restart CapRover?

**Comment by githubsaturn** [comment]
Yes! That should do it! And just let it soak for a month or two to make sure that the cets renewals are going through as well.

> Shame about the typo in the folder name letencrypt ;-)

I know!

**Comment by drmrbrewer** [comment]
OK I'll give it a go once I get my main instance running more smoothly (just posted an issue about service logs).

**Comment by drmrbrewer** [comment]
@githubsaturn 

It may be unrelated, but I tried deleting all files in those two folders, but when restarting CapRover I get the (unfortunately abbreviated) error message:

    # docker service update captain-captain --force
    captain-captain
    overall progress: 0 out of 1 tasks 
    1/1: invalid mount config for type "bind": bind source path does not exist: /ca… 
    service update paused: update paused due to failure or early termination of task p6sz50rsj3ef0r0sc50x4btp1

I get the same message (just a different task id) even after I copy the cert files back and restart CapRover as above.  Any ideas?  CapRover console does seem to restart OK despite the message.

**Comment by githubsaturn** [comment]
Shouldn't be related to the deleted files. You haven't deleted the directories, so it should be just fine. 
You can get more details by running `docker inspect TASK_ID`

**Comment by drmrbrewer** [comment]
extract:

        "Status": {
            "Timestamp": "2020-12-13T17:11:21.677752251Z",
            "State": "rejected",
            "Message": "preparing",
            "Err": "invalid mount config for type \"bind\": bind source path does not exist: /captain",
            "ContainerStatus": {
                "ContainerID": "",
                "PID": 0,
                "ExitCode": 0
            },
            "PortStatus": {}
        },

**Comment by githubsaturn** [comment]
Yea surely a Docker snafu... Try restarting docker and if it doesn't go away, maybe restart the machine.

**Comment by drmrbrewer** [comment]
> surely a Docker snafu

another one?  :-)  maybe not a big issue without restarting docker/machine, or is it necessary?  BTW even with the content of those folders gone, the world is still turning (so far).

**Comment by drmrbrewer** [comment]
Still working fine, though time will tell when the renewals start to go through.  It does seem that the `csr` and `keys` folders are gradually filling up with `.pem` files again (only up to 74 so far), but I've cleared them out again occasionally without ill effect.  I will also exclude these folder from the backup I take each night of the `/captain` folder.

**Comment by drmrbrewer** [comment]
To test how it would be if these files were cleared out automatically on a periodic basis, I have set up a crontab to run the following script every night:

    #!/bin/bash
    rm -f /captain/data/letencrypt/etc/csr/*
    rm -f /captain/data/letencrypt/etc/keys/*

**Comment by githubsaturn** [comment]
Yes, that should be fine. They responded to my request and said they won't be adding a prune method, instead, they are planning to "stop creating these files" https://github.com/certbot/certbot/issues/8525

**Comment by drmrbrewer** [comment]
@githubsaturn I notice that I have a `/captain/data/letencrypt/etc/archive` folder.  Within that are further folders, one each for a whole bunch of subdomains I've ever used in CapRover, and within those lots of `.pem` files.

Is this `/captain/data/letencrypt/etc/archive` folder something that is created and managed by CapRover?  Can I delete it or prune it like the other stuff mentioned in this thread?

**Comment by githubsaturn** [comment]
`.../archive` is managed by Let's Encrypt. You can view the details here: https://eff-certbot.readthedocs.io/en/stable/using.html#where-are-my-certificates

**Comment by drmrbrewer** [comment]
OK thanks, looks like it's safest to leave that one alone.

**Comment by TonyGravagno** [comment]
So ... it looks like this is still simply "open" but with no milestone. What is the intent for this issue? Over at certbot it looks like they've been sitting on tickets for years that would deprecate the backup files and eliminate the need for a prune. In the meantime it seems like the problem of junk files continues. It seems it would help if one side or the other took action? 🙏 Similar to @drmrbrewer, I've have a script to remove files older than a relative date. It would be nice to eliminate this some day.

TY

**Comment by githubsaturn** [comment]
Unless certbot has the capability to `prune` old certs, there is little that CapRover can do.
